Blog: Protect mission-critical pods with PriorityClasses (#37713)
kubernetes/website: content/en/blog/_posts/2023-01-12-protect-mission-critical-pods-priorityclass/index.md

---

**Author:** Sunny Bhambhani (InfraCloud Technologies)

Kubernetes has been widely adopted, and many organizations use it as their de-facto orchestration engine for running workloads that need to be created and deleted frequently.

Therefore, proper scheduling of the pods is key to ensuring that application pods are up and running within the Kubernetes cluster without any issues. This article delves into the use cases around resource management by leveraging the [PriorityClass](/docs/concepts/scheduling-eviction/pod-priority-preemption/#priorityclass) object to protect mission-critical or high-priority pods from getting evicted and making sure that the application pods are up, running, and serving traffic.

## Resource management in Kubernetes 

The control plane consists of multiple components, out of which the scheduler (usually the built-in [kube-scheduler](/docs/concepts/scheduling-eviction/kube-scheduler/)) is one of the components which is responsible for assigning a node to a pod.
- Whenever a pod is created, it goes to a pending state, and then the scheduler kicks in and determines which node should be the best-suited node for the placement of the new pod.

>   - In the background, the scheduler runs an infinite loop looking for pods with null in the nodeName field.

Reviewer: ```suggestion
  - In the background, the scheduler runs as an infinite loop looking for pods without a `nodeName` set
    (and that are [ready for scheduling](/docs/concepts/scheduling-eviction/pod-scheduling-readiness/)).
```

>   - When it finds one, it starts determining which node should be the best one for the pod.

Reviewer: ```suggestion
  - For each Pod that needs scheduling, the scheduler tries to make its best decision about which node should run that Pod.
```

>   - There are various parameters for determining the pod's placement, but to name a few, they are `nodeName` , `nodeSelector` , `taints & tolerations` , `nodeAffinity` , rank of available nodes based on available resources in terms of CPU and memory, etc.

Author: Does this sound good? because the rank of nodes is also one of the parameters that is decided by the scheduler, so I wanted to have that as one of the values.

To name a few, "node selector," "taints and tolerances," "node affinity," the rank of nodes based on available resources (for example, CPU and memory," and several other criteria are used to determine the pod's placement.

>   - There are various parameters for determining the pod's placement, but to name a few, they are `nodeName` , `nodeSelector` , `taints & tolerations` , `nodeAffinity` , rank of available nodes based on available resources in terms of CPU and memory, etc.

Author: Hi @sftim , 

Thank you for reviewing the article, I have made all the changes based on your recommendations.

Could you please have a quick check?

The preview version can be found here: https://deploy-preview-37713--kubernetes-io-main-staging.netlify.app/blog/2023/01/12/protect-mission-critical-pods-priorityclass/ 

PS: Also have corrected the typo in the image.

Thank you,
Sunny

- Just in case the scheduler is not able to find any node, the pod remains to be in the pending state, which is not an ideal scenario.

The below diagram explains how the request flows in (starting from point number 1 to 4):

> {{< figure src=kube-scheduler.svg alt="Whenever there is a new incoming pod, this is how kube-scheduler schedules a pod." title="Kube Scheduler">}}

Author: thanks, will correct it right away.

Just to add when you create a cluster using kubeadm, or even use a managed Kubernetes service (for example, Azure Kubernetes Service), Kubernetes uses PriorityClasses to safeguard the pods that are hosted on the controlplane nodes. This ensures that critical cluster components such as CoreDNS and kube-proxy are able to run even if resources are constrained.

It is achieved by using a special PriorityClass that ensures that these pods can serve the requests and that the overall cluster is up and running.

```console
$ kubectl get priorityclass
NAME                      VALUE        GLOBAL-DEFAULT   AGE
system-cluster-critical   2000000000   false            82m
system-node-critical      2000001000   false            82m
```

The diagram below shows exactly how it works with the help of an example, which will be detailed in the upcoming section.

{{< figure src=decision-tree.svg alt="Whenever there is a new incoming pod, this is how kube-scheduler decides their priority and puts them in a queue" title="Decision Tree">}}

### Pod priority and preemption

> [Pod preemption](/docs/concepts/scheduling-eviction/pod-priority-preemption/#preemption) is a Kubernetes feature that allows the cluster to preempt (evict) pods on the basis of priority and [Pod priority](/docs/concepts/scheduling-eviction/pod-priority-preemption/#pod-priority) indicates the importance of a pod relative to other pods while scheduling. If there aren't enough resources to run all the current pods, the scheduler tries to evict lower-priority pods over high-priority ones.

Reviewer: ```suggestion
[Pod preemption](/docs/concepts/scheduling-eviction/pod-priority-preemption/#preemption) is a Kubernetes feature that allows the cluster to preempt pods (removing an existing Pod in favor of a new Pod) on the basis of priority. [Pod priority](/docs/concepts/scheduling-eviction/pod-priority-preemption/#pod-priority) indicates the importance of a pod relative to other pods while scheduling. If there aren't enough resources to run all the current pods, the scheduler tries to evict lower-priority pods over high-priority ones.
```

You could talk either here or in a recap later in the article about how this ought to play out when a healthy cluster experiences a node failure: typically, lower priority Pods get preempted to create replacements for higher-priority Pods on the failed node. This happens even if the cluster is _also_ able to bring up a new node automatically, since Pod creation is usually much faster than node bringup.

Author: Sure, I'll go over the article again and include this scenario (perhaps I'll include it here itself since the reader may be in a flow and it might be helpful here itself).

4. Make sure that the pods for your workloads are running with the right PriorityClass.

## PriorityClass hands-on example

Let’s say there are 3 application pods: one for prod, one for preprod, and one for development. Below are three sample YAML manifest files for each of those.

```yaml
---
# development
apiVersion: v1
kind: Pod
metadata:
  name: dev-nginx
  labels:
    env: dev
spec:
  containers:
  - name: dev-nginx
    image: nginx
    resources:
      requests:
        memory: "256Mi"
         cpu: "0.2"
      limits:
        memory: ".5Gi"
        cpu: "0.5"
```

```yaml
---
# preproduction
apiVersion: v1
kind: Pod
metadata:
  name: preprod-nginx
  labels:
    env: preprod
spec:
  containers:
  - name: preprod-nginx
    image: nginx
    resources:
      requests:
        memory: "1.5Gi"
        cpu: "1.5"
      limits:
        memory: "2Gi"
        cpu: "2"
```

```yaml
---
# production
apiVersion: v1
kind: Pod
metadata:
  name: prod-nginx
  labels:
    env: prod
spec:
  containers:
  - name: prod-nginx
    image: nginx
    resources:
      requests:
        memory: "2Gi"
        cpu: "2"
      limits:
        memory: "2Gi"
        cpu: "2"
```

> Let's create these pods with the `kubectl create -f <FILE.yaml>` command and check their status by using the `kubectl get pods` command to see if they are up and running to serve the traffic.

Reviewer: ```suggestion
You can create these pods with the `kubectl create -f <FILE.yaml>` command, and then check their status
using the `kubectl get pods` command. You can see if they are up and look ready to serve traffic:
```

Let's see why this is happening:
```console
$ kubectl get events
...
...
5s          Warning   FailedScheduling   pod/prod-nginx      0/2 nodes are available: 1 Insufficient cpu, 2 Insufficient memory.
```

In this example, there is only one worker node and that node is having some resource crunch.

Now, let's look at how PriorityClass can help in this situation since prod should be given higher priority than the other environments.

## PriorityClass API

Before creating PriorityClasses based on these requirements, let's see what a basic manifest for a PriorityClass looks like and outline some prerequisites:

```yaml
apiVersion: scheduling.k8s.io/v1
kind: PriorityClass
metadata:
  name: PRIORITYCLASS_NAME
value: 0 # any integer value between -1000000000 to 1000000000 
description: >-
  (Optional) description goes here!
globalDefault: false # or true. Only one PriorityClass can be the global default.
```

Below are some prerequisites for PriorityClasses:

- The name of a PriorityClass must be a valid DNS subdomain name.

> - The name should not start with `system-`, as those names are reserved by Kubernetes (for example, they are used for two built-in  PriorityClasses).

Reviewer: ```suggestion
- When you make your own PriorityClass, the name should not start with `system-`, as those names are
  reserved by Kubernetes itself (for example, they are used for two built-in PriorityClasses).
```

- Its absolute value should be between -1000000000 to 1000000000 (1 billion).

> - Larger numbers are reserved by PriorityClasses such as system-cluster-critical, system-node-critical for kube-system Pods.

Reviewer: ```suggestion
- Larger numbers are reserved by PriorityClasses such as `system-cluster-critical`
  (this Pod is critically important to the cluster) and `system-node-critical` (the node
  critically relies on this Pod).
  `system-node-critical` is a higher priority than `system-cluster-critical`, because a
  cluster-critical Pod can only work well if the node where it is running has all its node-level
  critical requirements met.
```

- There are two optional fields:

>   - _globalDefault:_ This value is used for pods where the priorityClassName is not specified and only one PriorityClass with globalDefault set to true can exist in a cluster.  

Reviewer: ```suggestion
  - `globalDefault`: When true, this PriorityClass is used for pods where a `priorityClassName` is not specified.
    Only one PriorityClass with `globalDefault` set to true can exist in a cluster.  
```

  If there is no PriorityClass defined with globalDefault set to true, all the pods with no priorityClassName defined will be treated with 0 priority (i.e. the least priority).

>   - _Description:_ It's a string with a meaningful value so that people know when to use this PriorityClass.

Reviewer: ```suggestion
  - `description`:_ A string with a meaningful value so that people know when to use this PriorityClass.
```

{{< note >}}
Adding a PriorityClass with globalDefault set to true does not mean it will apply the same to the existing pods that are already running. This will be applicable only to the pods that came into existence after the PriorityClass was created.
{{< /note >}}

### PriorityClass in action

> Let's now create some environment-specific PriorityClasses:

Reviewer: ```suggestion
I'll explain with an example. I'll nowLet's now create some environment-specific PriorityClasses:
```
_or_
```suggestion
Here's an example. Next, create some environment-specific PriorityClasses:
```

The below diagram explains how the request flows in (starting from point number 1 to 4):

{{< figure src=kube-scheduler.svg alt="A diagram showing the scheduling of three Pods that a client has directly created." title="Scheduling in Kubernetes">}}

## Typical use cases

Below are some real-life scenarios where control over the scheduling and eviction of pods may be required.

1. Let's say the pod you plan to deploy is critical, and you have some resource constraints. An example would be the DaemonSet of an infrastructure component like Grafana Loki. The Loki pods must run before other pods can on every node. In such cases, you could ensure resource availability by manually identifying and deleting the pods that are not required or by adding a new node to the cluster. Both these approaches are unsuitable since the former would be tedious to execute, and the latter could involve an expenditure of time and money.

2. Another use case could be a single cluster that holds the pods for the below environments with associated priorities:
   - preprod [Priority 1] 
   - staging [Priority 2] 
   - development [Priority 3]

>     In the event of high resource consumption on the cluster, there is competition for CPU and memory resources on the nodes. While cluster-level autoscaling _may_ add more nodes, it takes time.  In the interim, if there are no further nodes to scale the cluster, some Pods could remain in a Pending state, or service could be degraded as they compete for resources. If the kubelet does evict a Pod from the node, that eviction would be random because the kubelet doesn’t have any special information about which Pods to evict and which to keep.

Reviewer 2: ```suggestion
 In the event of high resource consumption in the cluster, there is competition for CPU and memory resources on the nodes. While cluster-level autoscaling _may_ add more nodes, it takes time. In the interim, if there are no further nodes to scale the cluster, some Pods could remain in a Pending state, or the service could be degraded as they compete for resources. If the kubelet does evict a Pod from the node, that eviction would be random because the kubelet doesn’t have any special information about which Pods to evict and which to keep.
```

```console
$ kubectl get pods --show-labels
NAME            READY   STATUS    	RESTARTS   AGE   LABELS
dev-nginx       1/1     Terminating	0          55s   env=dev
preprod-nginx   1/1     Running   	0          55s   env=preprod
prod-nginx      0/1     Pending   	0          55s   env=prod
```

The dev-nginx pod is getting terminated. Once that is successfully terminated and there are enough resources for the prod pod, the control plane can schedule the prod pod:

```console
Warning   FailedScheduling   pod/prod-nginx    0/2 nodes are available: 1 Insufficient cpu, 2 Insufficient memory.
Normal    Preempted          pod/dev-nginx     by default/prod-nginx on node node01
Normal    Killing            pod/dev-nginx     Stopping container dev-nginx
Normal    Scheduled          pod/prod-nginx    Successfully assigned default/prod-nginx to node01
Normal    Pulling            pod/prod-nginx    Pulling image "nginx"
Normal    Pulled             pod/prod-nginx    Successfully pulled image "nginx"
Normal    Created            pod/prod-nginx    Created container prod-nginx
Normal    Started            pod/prod-nginx    Started container prod-nginx
```

## Summary

> The above example and its events clearly state what this feature of Kubernetes brings to the table, plus there can be multiple other scenarios as well where you can use this feature. To reiterate, this helps ensure that mission-critical pods are up and available to serve the traffic and, in the case of a resource crunch, determines cluster behavior.

Reviewer 2: ```suggestion
The above example and its events clearly state what this feature of Kubernetes brings to the table and multiple scenarios where you can use this feature. To reiterate, this helps ensure that mission-critical pods are up and available to serve the traffic and, in the case of a resource crunch, determines cluster behavior.
```

description: "Pod priority and preemption help to make sure that mission-critical pods are up in the event of a resource crunch by deciding order of scheduling and eviction."
---

**Author:** Sunny Bhambhani (InfraCloud Technologies)

Kubernetes has been widely adopted, and many organizations use it as their de-facto orchestration engine for running workloads that need to be created and deleted frequently.

Therefore, proper scheduling of the pods is key to ensuring that application pods are up and running within the Kubernetes cluster without any issues. This article delves into the use cases around resource management by leveraging the [PriorityClass](/docs/concepts/scheduling-eviction/pod-priority-preemption/#priorityclass) object to protect mission-critical or high-priority pods from getting evicted and making sure that the application pods are up, running, and serving traffic.

## Resource management in Kubernetes 

The control plane consists of multiple components, out of which the scheduler (usually the built-in [kube-scheduler](/docs/concepts/scheduling-eviction/kube-scheduler/)) is one of the components which is responsible for assigning a node to a pod.

> - Whenever a pod is created, it goes to a pending state, and then the scheduler kicks in and determines which node should be the best-suited node for the placement of the new pod.

Author: Its just about listing the scheduling steps, do let me know if you want me to use paragraph intead of bullet points.

> - Whenever a pod is created, it goes to a pending state, and then the scheduler kicks in and determines which node should be the best-suited node for the placement of the new pod.

Reviewer 2: A paragraph would be better because some of these steps are happening in parallel/in the background. A list, in my opinion, is for events occurring in a particular order/sequence.

> - Whenever a pod is created, it goes to a pending state, and then the scheduler kicks in and determines which node should be the best-suited node for the placement of the new pod.

Author: Updated:

`
In short, whenever a pod is created, it enters a "pending" state, after which the scheduler determines which node should be the best-suited node for the placement of the new pod.

In the background, the scheduler runs as an infinite loop looking for pods without a `nodeName` set and that are [ready for scheduling](/docs/concepts/scheduling-eviction/pod-scheduling-readiness/). For each Pod that needs scheduling, the scheduler tries to make its best decision about which node should run that Pod.

Just in case the scheduler is not able to find any node, the pod remains in the pending state, which is not an ideal scenario.

{{< note >}}
To name a few, `nodeSelector` , `taints and tolerations` , `nodeAffinity` , the rank of nodes based on available resources (for example, CPU and memory), and several other criteria are used to determine the pod's placement.
{{< note >}}
`

> - Whenever a pod is created, it goes to a pending state, and then the scheduler kicks in and determines which node should be the best-suited node for the placement of the new pod.

Reviewer 2: Whenever a pod is created, it enters a "pending" state, after which the scheduler determines which node is best suited for the placement of the new pod.

In the background, the scheduler runs as an infinite loop looking for pods without a nodeName set that are [ready for scheduling](https://github.com/docs/concepts/scheduling-eviction/pod-scheduling-readiness/). For each Pod that needs scheduling, the scheduler tries to decide which node should run that Pod.

If the scheduler cannot find any node, the pod remains in the pending state, which is not ideal.

{{< note >}}
To name a few, nodeSelector , taints and tolerations , nodeAffinity , the rank of nodes based on available resources (for example, CPU and memory), and several other criteria are used to determine the pod's placement.
{{< note >}}

> - Whenever a pod is created, it goes to a pending state, and then the scheduler kicks in and determines which node should be the best-suited node for the placement of the new pod.

Reviewer 2: Also, the figure below is getting rendered as a note.

> - Whenever a pod is created, it goes to a pending state, and then the scheduler kicks in and determines which node should be the best-suited node for the placement of the new pod.

Author: Corrected, missed to close note.

  - In the background, the scheduler runs as an infinite loop looking for pods without a `nodeName` set

>     (and that are [ready for scheduling](/docs/concepts/scheduling-eviction/pod-scheduling-readiness/)).

Author: This was added by @sftim and I believe the thought behined was to give a heads up about schedulingGates.
- https://kubernetes.io/docs/concepts/scheduling-eviction/pod-scheduling-readiness/

>     (and that are [ready for scheduling](/docs/concepts/scheduling-eviction/pod-scheduling-readiness/)).

Reviewer 2: I agree with the content, not the parentheses.

>     (and that are [ready for scheduling](/docs/concepts/scheduling-eviction/pod-scheduling-readiness/)).

Author: Removed parentheses.

  - For each Pod that needs scheduling, the scheduler tries to make its best decision about which node should run that Pod.
  - To name a few, `nodeSelector` , `taints and tolerations` , `nodeAffinity` , the rank of nodes based on available resources (for example, CPU and memory), and several other criteria are used to determine the pod's placement.  
- Just in case the scheduler is not able to find any node, the pod remains to be in the pending state, which is not an ideal scenario.

> The below diagram explains how the request flows in (starting from point number 1 to 4):

Author: Yes, we are talking about how to request flows in? and what all components are involved.

> The below diagram explains how the request flows in (starting from point number 1 to 4):

Reviewer 2: I don't think I made it clear. Are we talking about how a request makes its way into the system? Or are we talking about how it flows through the different components of the system.

> The below diagram explains how the request flows in (starting from point number 1 to 4):

Author: The intention was to have a very generic flow of how the request flows through different components and to have brief information about only components that are relevant to this topic.

> The below diagram explains how the request flows in (starting from point number 1 to 4):

Reviewer 2: ```suggestion
The below diagram, from point number 1 through 4, explains the request flow: 
```

{{< figure src=kube-scheduler.svg alt="A diagram showing the scheduling of three Pods that a client has directly created." title="Scheduling in Kubernetes">}}

## Typical use cases

Below are some real-life scenarios where control over the scheduling and eviction of pods may be required.

1. Let's say the pod you are planning to deploy is really critical, and you have some resource constraints. How will you make the resources available in the cluster? by first identifying which pods are not required, then manually deleting them, etc., which is a manual and tedious process, or you may plan to add a new node to the cluster, which too might take some time and money. 

   An example of such a case is a DaemonSet of an infrastructure component like Grafana Loki. In that case, the Loki pods must run on each node before any other pods run on that node.

2. Another use case can be: Suppose, you have a single cluster, and that cluster holds the pods for the below environments:
   - preprod [Priority 1] 
   - staging [Priority 2] 
   - development [Priority 3]

>    Unfortunately, the cluster goes through high resource consumption, and now there is a resource competition on the nodes with respect to the CPU and the memory. At least in the short term, you don’t have further nodes to expand or scale that Kubernetes cluster (cluster level autoscaling _may_ add more nodes, but this takes time). Perhaps some pods will remain in a Pending state, or perhaps service will degrade as Pods compete for resources. If the kubelet does evict a Pod from a node, that eviction happens at random because the kubelet doesn’t have any special information about which Pods to evict and which to keep.

Reviewer 2: ```suggestion
In the event of high resource consumption on the cluster, there is competition for CPU and memory resources on the nodes. While cluster-level autoscaling _may_ add more nodes, it takes time.  In the interim, if there are no further nodes to scale the cluster, some Pods could remain in a Pending state, or service could be degraded as they compete for resources. If the kubelet does evict a Pod from the node, that eviction would be random because the kubelet doesn’t have any special information about which Pods to evict and which to keep.
```

> 3. Assume you have a microservices architecture that includes Kafka, Zookeeper, MongoDB, a couple of jobs, a couple of CronJobs, and around 40+ microservices. Assume that if we are caught up in this situation of resource crunch and somehow Kafka or MongoDB gets evicted, then all our microservices will be useless until and unless these pods are up and start serving traffic again. 

Reviewer 2: This is a very specific microservice architecture. I'd recommend that we provide a generalized one and more context on why all microservices would be rendered useless if the Kafka/MongoDB pods get evicted. I know it sounds like I'm dumbing down things here, but readers may not have the context we have.

Author: Does this sounds good?

Assume you have a cloud-native application, i.e., a collection of small, independent, loosely coupled services and a couple of stateful applications like a queuing service or a database. And assume that if we are caught up in this situation of resource crunch and somehow the database pod gets evicted, then all our small services will be useless until and unless the database pod is up and start to serve traffic again.

Reviewer 2: Just rephrased your suggestion a bit.
```suggestion
3. A third example could be a microservice backed by a queuing application or a database running into a resource crunch and the queue or database getting evicted. In such a case, all the other services would be rendered useless until the database can serve traffic again.
```

Author: Looks good, Updated.

There can also be other scenarios where you want to control the order of scheduling or order of eviction of pods.

## PriorityClasses in Kubernetes

PriorityClass is a cluster-wide API object in Kubernetes and part of the `scheduling.k8s.io/v1` API group. It contains a mapping of the PriorityClass name (defined in `.metadata.name`) and an integer value (defined in `.value`). This represents the value that the scheduler uses to determine Pod's relative priority.

> Just to add when you create a cluster using kubeadm, or even use a managed Kubernetes service (for example, Azure Kubernetes Service), Kubernetes uses PriorityClasses to safeguard the pods that are hosted on the controlplane nodes. This ensures that critical cluster components such as CoreDNS and kube-proxy are able to run even if resources are constrained.

Reviewer 2: ```suggestion
Additionally, when you create a cluster using kubeadm or a managed Kubernetes service (for example, Azure Kubernetes Service), Kubernetes uses PriorityClasses to safeguard the pods that are hosted on the control plane nodes. This ensures that critical cluster components such as CoreDNS and kube-proxy can run even if resources are constrained.
```

> It is achieved by using a special PriorityClass that ensures that these pods can serve the requests and that the overall cluster is up and running.

Reviewer 2: What is achieved? And by who/what?

Author: Updated: 

This availability of pods is achieved through the use of a special PriorityClass that ensures the pods are up and running and that the overall cluster is not affected.

```console
$ kubectl get priorityclass
NAME                      VALUE        GLOBAL-DEFAULT   AGE
system-cluster-critical   2000000000   false            82m
system-node-critical      2000001000   false            82m
```

The diagram below shows exactly how it works with the help of an example, which will be detailed in the upcoming section.

{{< figure src="decision-tree.svg" alt="A flow chart that illustrates how the kube-scheduler prioritizes new Pods and potentially preempts existing Pods" title="Pod scheduling and preemption">}}

### Pod priority and preemption

[Pod preemption](/docs/concepts/scheduling-eviction/pod-priority-preemption/#preemption) is a Kubernetes feature that allows the cluster to preempt pods (removing an existing Pod in favor of a new Pod) on the basis of priority. [Pod priority](/docs/concepts/scheduling-eviction/pod-priority-preemption/#pod-priority) indicates the importance of a pod relative to other pods while scheduling. If there aren't enough resources to run all the current pods, the scheduler tries to evict lower-priority pods over high-priority ones.

> Also, when a healthy cluster experiences a node failure, typically, lower-priority pods get preempted to create room for higher-priority pods on the available node. This happens even if the cluster is also able to bring up a new node automatically, since pod creation is usually much faster than bringing up a new node.

Reviewer 2: ```suggestion
Also, when a healthy cluster experiences a node failure, typically, lower-priority pods get preempted to create room for higher-priority pods on the available node. This happens even if the cluster can bring up a new node automatically since pod creation is usually much faster than bringing up a new node.
```

### PriorityClass requirements

Before you set up PriorityClasses, there are a few things to consider.

1. Decide which PriorityClasses are needed. For instance, based on environment, type of pods, type of applications, etc.

> 2. Which PriorityClass resource will be the default for the cluster (otherwise, the pods without a `priorityClassName` will be treated as priority 0).

Reviewer 2: We should explain what priority 0 means here. Or link to it. 
```suggestion
2. The default PriorityClass resource for your cluster. The pods without a `priorityClassName` will be treated as priority 0.
```

        memory: "256Mi"
         cpu: "0.2"
      limits:
        memory: ".5Gi"
        cpu: "0.5"
```

```yaml
---
# preproduction
apiVersion: v1
kind: Pod
metadata:
  name: preprod-nginx
  labels:
    env: preprod
spec:
  containers:
  - name: preprod-nginx
    image: nginx
    resources:
      requests:
        memory: "1.5Gi"
        cpu: "1.5"
      limits:
        memory: "2Gi"
        cpu: "2"
```

```yaml
---
# production
apiVersion: v1
kind: Pod
metadata:
  name: prod-nginx
  labels:
    env: prod
spec:
  containers:
  - name: prod-nginx
    image: nginx
    resources:
      requests:
        memory: "2Gi"
        cpu: "2"
      limits:
        memory: "2Gi"
        cpu: "2"
```

You can create these pods with the `kubectl create -f <FILE.yaml>` command, and then check their status
using the `kubectl get pods` command. You can see if they are up and look ready to serve traffic:

```console
$ kubectl get pods --show-labels
NAME            READY   STATUS    RESTARTS   AGE   LABELS
dev-nginx       1/1     Running   0          55s   env=dev
preprod-nginx   1/1     Running   0          55s   env=preprod
prod-nginx      0/1     Pending   0          55s   env=prod
```

> Bad news. The pod for the Production environment is still Pending, which is bad since it won’t be serving any traffic. The production environment is the one where this matters most.

Reviewer 2: ```suggestion
Bad news. The pod for the Production environment is still Pending and isn't serving any traffic.
```

Let's see why this is happening:
```console
$ kubectl get events
...
...
5s          Warning   FailedScheduling   pod/prod-nginx      0/2 nodes are available: 1 Insufficient cpu, 2 Insufficient memory.
```

> In this example, there is only one worker node and that node is having some resource crunch.

Reviewer 2: ```suggestion
In this example, there is only one worker node, and that node has a resource crunch.
```

```console
$ kubectl get pods --show-labels
NAME            READY   STATUS    	RESTARTS   AGE   LABELS
dev-nginx       1/1     Terminating	0          55s   env=dev
preprod-nginx   1/1     Running   	0          55s   env=preprod
prod-nginx      0/1     Pending   	0          55s   env=prod
```

The dev-nginx pod is getting terminated. Once that is successfully terminated and there are enough resources for the prod pod, the control plane can schedule the prod pod:

```console
Warning   FailedScheduling   pod/prod-nginx    0/2 nodes are available: 1 Insufficient cpu, 2 Insufficient memory.
Normal    Preempted          pod/dev-nginx     by default/prod-nginx on node node01
Normal    Killing            pod/dev-nginx     Stopping container dev-nginx
Normal    Scheduled          pod/prod-nginx    Successfully assigned default/prod-nginx to node01
Normal    Pulling            pod/prod-nginx    Pulling image "nginx"
Normal    Pulled             pod/prod-nginx    Successfully pulled image "nginx"
Normal    Created            pod/prod-nginx    Created container prod-nginx
Normal    Started            pod/prod-nginx    Started container prod-nginx
```

## Summary

> The above example and its events clearly state what this feature of Kubernetes brings to the table, plus there can be multiple other scenarios as well where you can use this feature. Just to reiterate, this helps in making sure that mission-critical pods are made up and available to serve the traffic and, in the case of a resource crunch, how the cluster should behave.

Reviewer 2: ```suggestion
The above example and its events clearly state what this feature of Kubernetes brings to the table, plus there can be multiple other scenarios as well where you can use this feature. To reiterate, this helps ensure that mission-critical pods are up and available to serve the traffic and, in the case of a resource crunch, determines cluster behavior.
```

{{< /note >}}

The below diagram, from point number 1 through 4, explains the request flow: 

{{< figure src=kube-scheduler.svg alt="A diagram showing the scheduling of three Pods that a client has directly created." title="Scheduling in Kubernetes">}}

## Typical use cases

Below are some real-life scenarios where control over the scheduling and eviction of pods may be required.

1. Let's say the pod you plan to deploy is critical, and you have some resource constraints. An example would be the DaemonSet of an infrastructure component like Grafana Loki. The Loki pods must run before other pods can on every node. In such cases, you could ensure resource availability by manually identifying and deleting the pods that are not required or by adding a new node to the cluster. Both these approaches are unsuitable since the former would be tedious to execute, and the latter could involve an expenditure of time and money.

2. Another use case could be a single cluster that holds the pods for the below environments with associated priorities:
   - preprod [Priority 1] 
   - staging [Priority 2] 

>    - development [Priority 3]

Author: true that, thanks

- The name of a PriorityClass must be a valid DNS subdomain name.
- When you make your own PriorityClass, the name should not start with `system-`, as those names are
  reserved by Kubernetes itself (for example, they are used for two built-in PriorityClasses).
- Its absolute value should be between -1000000000 to 1000000000 (1 billion).
- Larger numbers are reserved by PriorityClasses such as `system-cluster-critical`
  (this Pod is critically important to the cluster) and `system-node-critical` (the node
  critically relies on this Pod).
  `system-node-critical` is a higher priority than `system-cluster-critical`, because a
  cluster-critical Pod can only work well if the node where it is running has all its node-level
  critical requirements met.
- There are two optional fields:
  - `globalDefault`: When true, this PriorityClass is used for pods where a `priorityClassName` is not specified.
    Only one PriorityClass with `globalDefault` set to true can exist in a cluster.  
  If there is no PriorityClass defined with globalDefault set to true, all the pods with no priorityClassName defined will be treated with 0 priority (i.e. the least priority).

>   - `description`:_ A string with a meaningful value so that people know when to use this PriorityClass.

Reviewer: ```suggestion
  - `description`: A string with a meaningful value so that people know when to use this PriorityClass.
```

```yaml
apiVersion: scheduling.k8s.io/v1
kind: PriorityClass
metadata:
  name: dev-pc
value: 1000000
globalDefault: false
description: >-
  (Optional) This priority class should only be used for all development pods.
```

```yaml
apiVersion: scheduling.k8s.io/v1
kind: PriorityClass
metadata:
  name: preprod-pc
value: 2000000
globalDefault: false
description: >-
  (Optional) This priority class should only be used for all preprod pods.
```

```yaml
apiVersion: scheduling.k8s.io/v1
kind: PriorityClass
metadata:
  name: prod-pc
value: 4000000
globalDefault: false
description: >-
  (Optional) This priority class should only be used for all prod pods.
```

Use `kubectl create -f <FILE.YAML>` command to create a pc and `kubectl get pc` to check its status.

```console
$ kubectl get pc
NAME                      VALUE        GLOBAL-DEFAULT   AGE
dev-pc                    1000000      false            3m13s
preprod-pc                2000000      false            2m3s
prod-pc                   4000000      false            7s
system-cluster-critical   2000000000   false            82m
system-node-critical      2000001000   false            82m
```

> The new PriorityClasses are in place now. A small change is needed in the pod manifest or pod template (in a ReplicaSet or Deployment). In other words, we need to specify the priority class name at `.spec.priorityClassName` (which is a string value).

Reviewer: ```suggestion
The new PriorityClasses are in place now. A small change is needed in the pod manifest or pod template (in a ReplicaSet or Deployment). In other words, you need to specify the priority class name at `.spec.priorityClassName` (which is a string value).
```

First update the previous production pod manifest file to have a PriorityClass assigned, then delete the Production pod and recreate it. You can't edit the priority class for a Pod that already exists.

> That change seems successful; the status of pods has been updated:

Reviewer: (if you want to)
```suggestion
In my cluster, when I tried this, here's what happened.
First, that change seems successful; the status of pods has been updated:
```

```console
$ kubectl get pods --show-labels
NAME            READY   STATUS    	RESTARTS   AGE   LABELS
dev-nginx       1/1     Terminating	0          55s   env=dev
preprod-nginx   1/1     Running   	0          55s   env=preprod
prod-nginx      0/1     Pending   	0          55s   env=prod
```

The dev-nginx pod is getting terminated. Once that is successfully terminated and there are enough resources for the prod pod, the control plane can schedule the prod pod:

```console
Warning   FailedScheduling   pod/prod-nginx    0/2 nodes are available: 1 Insufficient cpu, 2 Insufficient memory.
Normal    Preempted          pod/dev-nginx     by default/prod-nginx on node node01
Normal    Killing            pod/dev-nginx     Stopping container dev-nginx
Normal    Scheduled          pod/prod-nginx    Successfully assigned default/prod-nginx to node01
Normal    Pulling            pod/prod-nginx    Pulling image "nginx"
Normal    Pulled             pod/prod-nginx    Successfully pulled image "nginx"
Normal    Created            pod/prod-nginx    Created container prod-nginx
Normal    Started            pod/prod-nginx    Started container prod-nginx
```

> 

Reviewer: Readers might assume that these PriorityClasses are somehow enforced, which isn't true.

You could add something similar to:
```suggestion

## Enforcement

When you set up PriorityClasses, they exist just how you defined them. However, people
(and tools) that make changes to your cluster are free to set any PriorityClass, or to not
set any PriorityClass at all.
However, you can use other Kubernetes features to make sure that the priorities you wanted
are actually applied.

As an alpha feature, you can define a [ValidatingAdmissionPolicy](/blog/2022/12/20/validating-admission-policies-alpha/) and a ValidatingAdmissionPolicyBinding so that, for example,
Pods that go into the `prod` namespace must use the `prod-pc` PriorityClass.
With another ValidatingAdmissionPolicyBinding you ensure that the `preprod` namespace
uses the `preprod-pc` PriorityClass, and so on.
In *any* cluster, you can enforce similar controls using external projects such as
[Kyverno](https://kyverno.io/) or [Gatekeeper](https://open-policy-agent.github.io/gatekeeper/),
through validating admission webhooks.

However you do it, Kubernetes gives you options to make sure that the PriorityClasses are
used how you wanted them to be, or perhaps just to
[warn](https://open-policy-agent.github.io/gatekeeper/website/docs/violations/#warn-enforcement-action)
users when they pick an unsuitable option.

```

```

Author: This is a nice piece of information, though I haven't tried it yet, I will surely give it a try.

Since this feature is in alpha, Maybe in the future we can create a part-2 on this and can link that with this article. What do you think? 

Also about the enforcement part, yes true we can add below section.

## Enforcement
When you set up PriorityClasses, they exist just how you defined them. However, people (and tools) that make changes to your cluster are free to set any PriorityClass, or not to set any PriorityClass at all.

Reviewer: ValidatingAdmissionPolicy is alpha, but [admission webhooks](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#what-are-admission-webhooks) are stable. It's important to cover the overall admission control mechanism at least in outline, because it is relevant.

If you want to explain that aspect in another article, turning this into 2 parts, that's absolutely fine -  but I'd want to see a PR opened and drafted for part 2 before we merge the PR for part 1.

If you do opt to add a second part, that gives you room to cover some of the details from https://kubernetes.io/docs/concepts/policy/resource-quotas/#resource-quota-per-priorityclass

---

The minimum here is to mention that admission control is often an important part of making priorityclasses work right; you must mention that admission control concept somehow. Maybe in only 50 more words.

If you want to go beyond that, you have options. I'm fine with making this a 2 part piece, even if that means a delay. It's your article, so you do get to choose which option you prefer.

Author: That makes sense, I will go with your suggestion. 

But still maybe in near future, I will write an article about ValidatingAdmissionPolicy and resource-quota-per-priorityclass for sure.

Thank you for all your suggestions and time.

## Summary

The above example and its events clearly state what this feature of Kubernetes brings to the table and multiple scenarios where you can use this feature. To reiterate, this helps ensure that mission-critical pods are up and available to serve the traffic and, in the case of a resource crunch, determines cluster behavior.

It gives you some power to decide the order of scheduling and order of [preemption](/docs/concepts/scheduling-eviction/pod-priority-preemption/#preemption) for Pods. Therefore, you need to define the PriorityClasses sensibly.

> 

Reviewer: ```suggestion
For example, if you have a cluster autoscaler to add nodes on demand,
make sure to run it with the `system-cluster-critical` PriorityClass. You don't want to
get in a situation where the autoscaler has been preempted and there are no new nodes
coming online.

```
?
